flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/table.scala

```diff
@@ -999,6 +999,93 @@ class Table(
     new OverWindowedTable(this, overWindows.toArray)
   }
 
+  /**
+    * Performs a map operation with an user-defined function. The output will be flattened
+    * if the output type is composite type.
+    *
+    * Scala Example:
+    * {{{
+    *   class MyMapFunction extends ScalarFunction {
+    *     def eval(str: String): Row = {
+    *       Row.of(str, str)
+    *     }
+    *
+    *     def getResultType(signature: Array[Class[_]]): TypeInformation[_] =
+    *       Types.ROW(Types.STRING, Types.STRING)
+    *   }
+    *
+    *   val func = new MyMapFunction()
+    *   table.map(func('c))
+    * }}}
+    *
+    * Java Example:
+    * {{{
+    *   class MyMapFunction extends ScalarFunction {
+    *     public Row eval(String str) {
+    *       return Row.of(str, str);
+    *     }
+    *
+    *     public TypeInformation getResultType(Class[] signature) {
+    *       return new RowTypeInfo(BasicTypeInfo.STRING_TYPE_INFO, BasicTypeInfo.STRING_TYPE_INFO);
+    *     }
+    *   }
+    *
+    *   ScalarFunction func = new MyMapFunction();
+    *   tableEnv.registerFunction("func", func);
+    *   table.map("func(c)");
+    * }}}
+    */
+  def map(mapFunction: Expression): Table = {
+    mapFunction match {
+      case _: ScalarFunctionCall =>
+      case _ => throw new ValidationException("Only ScalarFunction can be used in map.")
+    }
+
+    val expandedFields = expandProjectList(Seq(Flattening(mapFunction)), logicalPlan, tableEnv)
```

Comment:
we should not flatten the output if the result type for scalar function is not a composite type.

Reply:
It's safe to always flatten the result even if it is not a composite type.

Reply:
Here you use flatten to implement map, map needs to guarantee non-deterministic and deterministic user-defined scalar function works well, otherwise the reason must be explained,What do you think?

Reply:
+1